src/types/string.rs

```diff
@@ -165,61 +169,40 @@ impl PyString {
     pub fn from_object<'p>(src: &'p PyAny, encoding: &str, errors: &str) -> PyResult<&'p PyString> {
         unsafe {
             src.py()
-                .from_owned_ptr_or_err::<PyString>(ffi::PyUnicode_FromEncodedObject(
+                .from_owned_ptr_or_err(ffi::PyUnicode_FromEncodedObject(
                     src.as_ptr(),
                     encoding.as_ptr() as *const c_char,
                     errors.as_ptr() as *const c_char,
                 ))
         }
     }
 
-    /// Gets the Python string as a byte slice.
+    /// Gets the Python string as a Rust UTF-8 string slice.
     ///
     /// Returns a `UnicodeEncodeError` if the input is not valid unicode
     /// (containing unpaired surrogates).
-    #[inline]
     pub fn to_str(&self) -> PyResult<&str> {
-        let utf8_slice = {
-            cfg_if::cfg_if! {
-                if #[cfg(any(Py_3_10, not(Py_LIMITED_API)))] {
-                    // PyUnicode_AsUTF8AndSize only available on limited API starting with 3.10.
-                    let mut size: ffi::Py_ssize_t = 0;
-                    let data: *const u8 = unsafe { ffi::PyUnicode_AsUTF8AndSize(self.as_ptr(), &mut size).cast() };
-                    if data.is_null() {
-                        return Err(crate::PyErr::fetch(self.py()));
-                    } else {
-                        unsafe { std::slice::from_raw_parts(data, size as usize) }
-                    }
-                } else {
-                    let bytes = unsafe {
-                        self.py().from_owned_ptr_or_err::<PyBytes>(ffi::PyUnicode_AsUTF8String(self.as_ptr()))?
-                    };
-                    bytes.as_bytes()
-                }
-            }
-        };
-        Ok(unsafe { std::str::from_utf8_unchecked(utf8_slice) })
+        #[cfg(any(Py_3_10, not(Py_LIMITED_API)))]
+        {
+            Py2Borrowed::from_gil_ref(self).to_str()
+        }
+
+        #[cfg(not(any(Py_3_10, not(Py_LIMITED_API))))]
+        {
+            let bytes = unsafe {
+                self.py()
+                    .from_owned_ptr_or_err::<PyBytes>(ffi::PyUnicode_AsUTF8String(self.as_ptr()))
+            }?;
+            Ok(unsafe { std::str::from_utf8_unchecked(bytes.as_bytes()) })
+        }
     }
 
     /// Converts the `PyString` into a Rust string.
     ///
     /// Unpaired surrogates invalid UTF-8 sequences are
     /// replaced with `U+FFFD REPLACEMENT CHARACTER`.
     pub fn to_string_lossy(&self) -> Cow<'_, str> {
```

Comment:
Maybe it would help migrating downstream code if `PyString` also had the `to_cow` method, so that using that method and changing the pointer type could be separate steps.

Reply:
Good idea, added.